### PR TITLE
CI: Fix and restore PySpark and Spark tests

### PR DESCRIPTION
**Pull request by timothydijamco**
This is a continuation of #2212 (again the focus is on fixing and restoring _PySpark_ tests, but not the _Spark_ tests).

I'm expecting the CI to fail here at first because of some PySpark tests, will be continuing with more changes to fix them

**edit:** Originally, this PR was for fixing only the PySpark tests, but both can be fixed in the same way, so now fixing both in this PR

### CODE REVIEW

**Comment by timothydijamco** [comment]
With `485adc5` (`Pin pyarrow to be between 0.13 and 0.14.1`), the `LinuxTest py37_omnisci_spark` check should succeed (but take a while for Conda to solve) and the `LinuxTest py36_omnisci_spark` check should fail due to requirements conflicts (at least this is what I saw locally)

File: ci/requirements-3.7-dev.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ dependencies:
   - plumbum
   - pre-commit
   - psycopg2
-  - pyarrow>=0.13
+  - pyarrow>=0.13, <=0.14.1
```
Comment (jreback) [inline ci/requirements-3.7-dev.yml:28]: really? why do we actually need this? i would rather move up the pyarrow >= 0.14 (or even higher)

Comment (timothydijamco) [reply ci/requirements-3.7-dev.yml:28]: I still don't know specifically why, but some tests in `test_temporal.py` fail if pyarrow is 0.15.0 higher and succeed if pyarrow is 0.14.1 or lower.

pyarrow used to happen to solve to 0.13.0, but unrelated changes in requirements.yml ([in this commit](https://github.com/ibis-project/ibis/commit/3d9756cee1dc3e0c622891dd6314c2f0abecc382) inadvertently made pyarrow solve to 0.17.1 (if `pymapd` is commented out from the requirements yml), or to 0.15.0 (if `pymapd` is enabled in the requirements yml)

I could still try separate requirements .yml files for each backend set (would have to split omnisci and PySpark apart because of the requirements conflicts). But it sounds like maybe it's worth it for me to see if the failing tests themselves can be changed so we don't have to limit the version

Comment (jreback) [reply ci/requirements-3.7-dev.yml:28]: so i am ok on *some* builds to use older versions, but what I would suggest is that we create a new .yml for *just* spark (you will have to separate it from the omnisci backend) and just go from there to get this working.

Comment (icexelloss) [reply ci/requirements-3.7-dev.yml:28]: The newest Spark release (2.4) depends on Arrow 0.10 on the java side. It's possible that it's too old and doesn't work with newer version of PyArrow.

Comment (jreback) [reply ci/requirements-3.7-dev.yml:28]: ok fair enough. let's get this working, then create an issue about using the latest versions (or simply change the supported versions).

**Comment by timothydijamco** [comment]
The CI passes now

The fix to get the PySpark tests working again also fixes the Spark tests, so I've re-enabled them as well. I'm going to rename this PR to include both PySpark and Spark

File: ci/requirements-3.6-dev-pyspark-spark.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  - multipledispatch>=0.6.0
+  - mypy
+  - numpy>=1.15
+  - openjdk=8
+  - pandas>=0.25.3
+  - pip=19.3.1
+  - plumbum
+  - pre-commit
+  - psycopg2
+  - pyarrow>=0.13, <=0.14.1
```
Comment (timothydijamco) [inline ci/requirements-3.6-dev-pyspark-spark.yml:28]: These new `requirements-3.x-dev-pyspark-spark.yml` files are the same as `requirements-3.x-dev.yml` except `pymapd` is disabled and `<=0.14.1` is added to the `pyarrow` version pin

Comment (icexelloss) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: Does it make sense to remove all the non spark dependency here? e.g., pymysql, impyla, bigquery, clickhouse, etc? I think

Comment (icexelloss) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: I think to run Pyspark CI you actually only need very minimal dependencies, i.e., pyspark, pyarrow, pandas and numpy

File: ci/azure/linux.yml [hunk 3]
```diff
@@ -155,7 +173,7 @@ jobs:
     - task: PublishPipelineArtifact@1
       inputs:
         path: /tmp/env/env.yml
-        artifact: LinuxCondaEnvironment$(PYTHON_NO_DOT_VERSION)
+        artifact: LinuxCondaEnvironment-$(PIPELINE_ARTIFACT_TAG)
```
Comment (icexelloss) [inline ci/azure/linux.yml:176]: Why this change too?

File: ci/azure/linux.yml [hunk 4]
```diff
@@ -81,7 +96,10 @@ jobs:
       displayName: 'Setup BigQuery credentials'
       condition: eq(variables['System.PullRequest.IsFork'], 'False')
 
-    - bash: make start PYTHON_VERSION=$PYTHON_VERSION BACKENDS="${BACKENDS}"
+    - bash: |
+        if [ -n "${BACKENDS}" ]; then
```
Comment (icexelloss) [inline ci/azure/linux.yml:100]: Why the if statement?

File: ci/azure/linux.yml (hunk 3 above)
Comment (timothydijamco) [reply ci/azure/linux.yml:176]: The `LinuxTest py3X_pyspark_spark` jobs would fail on this step because they are trying to publish an artifact with the same name as an existing artifact

Our existing jobs already do publish artifacts with duplicate names completely fine, but I'm guessing the error comes up specifically for the `_pyspark_spark` jobs because the artifact is different for those jobs (contents of `/tmp/env/env.yml` is different, see line 175, line 135). So the jobs are trying to publish that different artifact to an existing name

E.g., this is what would happen before this change:

1. `LinuxTest py36_omnisci` publishes artifact `LinuxCondaEnvironment36`
2. `LinuxTest py36_sql_parquet` publishes artifact `LinuxCondaEnvironment36` (no error, possibly because the contents of the artifact input `/tmp/env/env.yml` is the same as above?)
3. `LinuxTest py36_pyspark_spark` tries to publish artifact `LinuxCondaEnvironment36` (with an artifact input `/tmp/env/env.yml` that has different contents) and fails with this error: `Artifact LinuxCondaEnvironment36 already exists for build 2515.`

File: ci/azure/linux.yml (hunk 4 above)
Comment (timothydijamco) [reply ci/azure/linux.yml:100]: `make start` with `BACKENDS=""` starts all the backends instead of none of them because it essentially ends up running ```docker-compose up  ```  which starts all services (compared to something like ```docker-compose up mysql``` which would only start the mysql service)

File: ci/requirements-3.6-dev-pyspark-spark.yml (hunk 2 above)
Comment (timothydijamco) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: That makes sense to me, I'll do that

Comment (timothydijamco) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: Hopefully reduces the Conda solve time too, it's highest in the PySpark/Spark CI jobs

File: ci/azure/linux.yml (hunk 4 above)
Comment (jreback) [reply ci/azure/linux.yml:100]: can you add a comment to this effect

File: ci/Dockerfile.dev [hunk 5]
```diff
@@ -6,14 +6,15 @@ RUN apt-get -qq update --yes \
  && rm -rf /var/lib/apt/lists/*
 
 ARG PYTHON_VERSION
-ADD ci/requirements-$PYTHON_VERSION-dev.yml /
+ARG REQUIREMENTS_TAG
```
Comment (jreback) [inline ci/Dockerfile.dev:9]: i think we have an explanation of these parms somewhere can you find and update docs

File: ci/requirements-3.6-dev-pyspark-spark.yml (hunk 2 above)
Comment (jreback) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: can you add a comment / issue link about this (do we have an open ibis issue for this?)

File: ci/docker-compose.yml [hunk 6]
```diff
@@ -111,6 +111,7 @@ services:
       dockerfile: ci/Dockerfile.dev
       args:
         PYTHON_VERSION: ${PYTHON_VERSION:-3.6}
+        REQUIREMENTS_TAG: ${REQUIREMENTS_TAG}
```
Comment (jreback) [inline ci/docker-compose.yml:114]: do the other backends now have there own  requirements-tag? if not I think just change all of them to have that

File: ci/requirements-3.6-dev-pyspark-spark.yml (hunk 2 above)
Comment (timothydijamco) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: Do you mean a comment/issue about the solve time? I think in general the solve time isn't really an issue now after [this commit](https://github.com/ibis-project/ibis/commit/3d9756cee1dc3e0c622891dd6314c2f0abecc382)—I just noticed that Conda takes a while to solve in the new PySpark/Spark CI jobs I'm adding here. Removing the extra dependencies would speed it up though and make the CI finish a bit faster

Comment (timothydijamco) [reply ci/requirements-3.6-dev-pyspark-spark.yml:28]: Oh I just saw you commented this also on the other requirements file, do you mean create an issue about the pyarrow version (increasing it, or changing supported versions)?

File: ci/docker-compose.yml (hunk 6 above)
Comment (timothydijamco) [reply ci/docker-compose.yml:114]: I can do that, I'll probably make the requirements tag `main` for the other backends to make it concise:

- SQL, Parquet, Impala, Kudu, Clickhouse, Omnisci
  - Requirements tag will be `main`
  - Requirements file will be named `requirements-3.x-dev-main.yml`
- PySpark, Spark: 
  - Requirements tag is `pyspark-spark`
  - Requirements file is named `requirements-3.x-dev-pyspark-spark.yml`

(I want to keep `dev` in the filename because other parts of the CI have a distinction between `docs` (for the documentation) and `dev` (for the code))

File: Makefile [hunk 7]
```diff
@@ -3,16 +3,18 @@
...
 MAKEFILE_DIR = $(patsubst %/,%,$(dir $(abspath $(lastword $(MAKEFILE_LIST)))))
 
-# from PYTHON_VERSION depends which `./ci/requirements-$PYTHON_VERSION-dev` file
-# will be used for creating ibis image (see for additional info: `./ci/Dockerfile.dev`
-# and `./ci/docker-compose.yml`)
-# you can use `3.6` or `3.7` for now
+# PYTHON_VERSION and REQUIREMENTS_TAG defines which `./ci/requirements-dev-$PYTHON_VERSION-$REQUIREMENTS_TAG`
+# file will be used for creating the ibis image (see for additional info: `./ci/Dockerfile.dev` and
+# `./ci/docker-compose.yml`)
+# You can use `3.6` or `3.7` for now for the PYTHON_VERSION
```
Comment (jreback) [inline Makefile:9]: next pr we can add 3.8 to this (not worth it here)

**Comment by jreback** [comment]
thanks @timothydijamco very nice!

**Comment by timothydijamco** [comment]
thank you!
